pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -1064,3 +1064,65 @@ func setUpstreamProtocol(node *model.Proxy, c *cluster.Cluster, port *model.Port
 		c.ProtocolSelection = cluster.Cluster_USE_DOWNSTREAM_PROTOCOL
 	}
 }
+
+func addTelemetryMetadata(opts buildClusterOpts) {
+	if opts.cluster == nil || opts.proxy == nil || opts.proxy.ServiceInstances == nil ||
+		len(opts.proxy.ServiceInstances) == 0 {
+		return
+	}
+	if opts.cluster.Metadata == nil {
+		opts.cluster.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*structpb.Struct{},
+		}
+	}
+
+	// Create Istio metadata if does not exist yet
+	if _, ok := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]; !ok {
+		opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey] = &structpb.Struct{
+			Fields: map[string]*structpb.Value{},
+		}
+	}
+
+	im := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]
+
+	// If services already exist in cluster metadata, delete it and recreate the field,
+	// since service metadata should only be added only in this function.
+	delete(im.Fields, "services")
```

Comment:
nit: L1093 does the same thing anyways since it will overwrite it?

Reply:
Ah yes. thanks!